doc/tutorial/tutorial.md

```diff
@@ -1,40 +1,650 @@
+`subnet_id`, `subnet_prefix`, `subnet_mask` variables to records from `Host` and `Subnet` relations.
+
+The first two clauses iterate over all host-subnet pairs.  We would like to narrow this set down to
+only those pairs where the host's IP address belongs to the subnet.  This is captured by the
+expression in the third clause of the rule: `(host_ip & subnet_mask) == subnet_prefix`, where `&` is
+the bitwise "and" operaror, and `==` is the equality operator.  In DDlog, a clause that is a Boolean
+expression is a *filter* clause that eliminates all tuples that do not satisfy the condition.
+
+Note that this rule does not use the `name` field of the `Host` relation.  We do not want to pollute
+the name space with an unused variable and instead use the *wildcard* symbol (`_`) in the first
+clause of the rule to indicate that the value of the field is immaterial:
+
+```
+Host(host_id, _, host_ip)
+```
+
+DDlog supports an alternative syntax for atoms, identifying relation arguments by name instead of by
+position:
+
+```
+Host(.id=host_id, .ip=host_ip)
+```
+
+While this syntax is more verbose, it improves code clarity and prevents errors due to incorrect
+ordering of arguments.  In addition, it only requires the programmer to list arguments that are used
+in the rule.  Omitted arguments are automaticaly wildcarded.
+
+### String literals
+
+DDlog has built-in support for strings, allowing many string-manipulating programs to be expressed
+completely within the language, without relying on external functions.  DDlog's primitive `string`
+type contains UTF-8 strings.  Two forms of string literals are supported:
+
+1. *Quoted strings* are Unicode strings enclosed in quotes, with the same single-character
+escapes as in C, e.g.,
+  ```
+  "bar", "\tbar\n", "\"buzz", "\\buzz", "barΔ"
+  ```
+  A quoted string may not contain an unescaped new-line character, backslash, or quote, e.g.,
+  ```
+  "foo
+  bar"
+  ```
+  is illegal.  Long strings can be broken into multiple lines using backslash:
+  ```
+  "foo\
+     \bar"
+  ```
+  is equivalent to `"foobar"`
+
+1. *Raw strings*, enclosed in `[| |]`, can contain arbitrary Unicode characters, except the `|]`
+sequence, including newlines and backslashes, e.g.,
+  ```
+  [|
+  foo\n
+  buzz
+  bar|]
+  ```
+  is the same string as `"foo\\n\nbuzz\nbar"`.
+
+Adjacent string literals (of both kinds) are automatically concatenated, e.g., `"foo" "bar"` is
+converted to `"foobar"`.
+
+See [`strings.dl`](../../test/datalog_tests/strings.dl) for more examples of string literals.
 
 ### String concatenation and interpolation
 
-### Arithmetic expressions
+We have already encountered the string concatenation operator `++`.  Another way to construct
+strings in DDlog is using *string interpolation*, which allows embedding arbitrary DDlog expressions
+inside a string literal.  At runtime, the result of the expression is automatically converted to a
```

Comment:
expressions are evaluated, converted to strings and the result is inserted within the string.